### PR TITLE
Improve BWC distribution mismatch error message

### PR DESCRIPTION
**Pull request by mark-vieira**
After version bumps, it's very common for open PR branches to start failing because of the missing version constants breaking BWC distribution assumptions. You end up getting  somewhat obscure error that doesn't give you much indication as to the _reason_ things failed:

```
* What went wrong:
Execution failed for task ':distribution:bwc:bugfix:buildBwcLinuxTar'.
> Building 8.3.3 didn't generate expected artifact /dev/shm/elastic+elasticsearch+pull-request+bwc/distribution/bwc/bugfix/build/bwc/checkout-8.3/distribution/archives/linux-tar/build/install/elasticsearch-8.3.3-SNAPSHOT
```

This PR amends this error message to give developers some hint as to what might be wrong:

```
* What went wrong:
Execution failed for task ':distribution:bwc:bugfix:buildBwcLinuxTar'.
> Building 8.3.3 didn't generate expected artifact [distribution/bwc/bugfix/build/bwc/checkout-8.3/distribution/archives/linux-tar/build/install/elasticsearch-8.3.3-SNAPSHOT]. The working branch may be out of date.
```

### CODE REVIEW

File: build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/InternalDistributionBwcSetupPlugin.java [hunk 1]
```diff
@@ -249,8 +250,14 @@ static void createBuildBwcTask(
...
                     if (expectedOutputFile.exists() == false) {
+                        Path relativeOutputPath = project.getRootDir().toPath().relativize(expectedOutputFile.toPath());
                         throw new InvalidUserDataException(
-                            "Building " + bwcVersion.get() + " didn't generate expected artifact " + expectedOutputFile
+                            "Building "
+                                + bwcVersion.get()
+                                + " didn't generate expected artifact ["
+                                + relativeOutputPath
+                                + "]. "
+                                + "The working branch may be out of date."
```
Comment (pugnascotia) [inline build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/InternalDistributionBwcSetupPlugin.java:260]:
```suggestion
                            "Building "
                                + bwcVersion.get()
                                + " didn't generate expected artifact ["
                                + relativeOutputPath
                                + "]. "
                                + "The working branch may be out of date."
```

File: build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/InternalDistributionBwcSetupPlugin.java [hunk 2]
```diff
@@ -249,8 +250,14 @@ static void createBuildBwcTask(
                 @Override
                 public void execute(Task task) {
                     if (expectedOutputFile.exists() == false) {
+                        Path relativeOutputPath = project.getRootDir().toPath().relativize(expectedOutputFile.toPath());
                         throw new InvalidUserDataException(
-                            "Building " + bwcVersion.get() + " didn't generate expected artifact " + expectedOutputFile
+                            "Building "
```
Comment (pugnascotia) [inline build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/InternalDistributionBwcSetupPlugin.java:255]: Maybe we can go further and explicitly suggest a merge?

```java
final String message = "Building %s didn't generate expected artifact [%s]. The working branch may be "
    + "out-of-date - try merging in the latest upstream changes to the branch.";
throw new InvalidUserDataException(message.formatted(bwcVersion.get(), relativeOutputPath));
```

**Review by pugnascotia — APPROVED** [review]

File: build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/InternalDistributionBwcSetupPlugin.java (hunk 2 above)
Comment (mark-vieira) [reply build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/InternalDistributionBwcSetupPlugin.java]: Yes, that is indeed better.

**Comment by mark-vieira** [comment]
@elasticmachine update branch

**Pull request merged**